R/Data-validity.R

```diff
@@ -1,114 +1,100 @@
-#' valid elements with respect to their types, dependency and length.
-#'
+#' @describeIn validate_data_objects validates that the [`Data`] object contains
+#'   valid elements with respect to their types, dependency and length.
 validate_data <- function(object) {
   o <- Validate()
-
   o$check(
-    test_double(object@x, len = object@nObs),
+    test_double(object@x, len = object@nObs, any.missing = FALSE),
     "Doses vector x must be of type double and length nObs"
   )
-
   o$check(
-    test_integer(object@y, lower = 0, upper = 1, len = object@nObs),
-    "DLT vector y must be of type integer of lenght nObs and can only take 0 or 1 values"
+    test_integer(object@y, lower = 0, upper = 1, len = object@nObs, any.missing = FALSE),
+    "DLT vector y must be nObs long and contain 0 or 1 integers only"
   )
-
   o$check(
-    test_double(object@doseGrid, len = object@nGrid),
-    "Possible doses vector doseGrid must be of type double and length nGrid"
+    test_double(object@doseGrid, len = object@nGrid, any.missing = FALSE, sorted = TRUE, unique = TRUE),
+    "doseGrid must be of type double and length nGrid and contain unique, sorted values"
   )
-
   o$check(
     test_int(object@nGrid),
-    "Number of gridpoints nGrid must be of type integer"
+    "Number of dose grid values nGrid must be scalar integer"
   )
-
   o$check(
-    test_integer(object@xLevel, len = object@nObs),
-    "Levels xLevel for the doses the patients have been given must be of type integer
-     and length nObs"
+    test_integer(object@xLevel, len = object@nObs, any.missing = FALSE),
+    "Levels xLevel for the doses the patients have been given must be of type integer and length nObs"
   )
-
-  o$check(test_flag(object@placebo), "The placebo flag must be of type logical")
-
   o$check(
-    !is.unsorted(object@doseGrid, strictly = TRUE),
-    "doseGrid must be sorted and without duplicate values"
+    test_flag(object@placebo), 
+    "The placebo flag must be scalar logical"
   )
-
   o$check(
     test_subset(object@x, object@doseGrid),
     "Dose values in x must be from doseGrid"
   )
-
   o$check(
-    all_equal(object@x, object@doseGrid[object@xLevel]),
-    "x must be equal to doseGrid[xLevel] (tolerance 1e-10)"
+    h_all_equivalent(object@x, object@doseGrid[object@xLevel]),
+    "x must be equivalent to doseGrid[xLevel] (up to numerical tolerance)"
   )
-
   if (object@placebo) {
-    plcb <- object@x == object@doseGrid[1] # indicator of placebo
+    is_placebo <- object@x == object@doseGrid[1]
     o$check(
-      test_set_equal(object@cohort, object@cohort[!plcb]),
+      test_set_equal(object@cohort, object@cohort[!is_placebo]),
       "A cohort with only placebo is not allowed"
     )
-  } else {
-    plcb <- FALSE
+    o$check(
+      h_doses_unique_per_cohort(dose = object@x[!is_placebo], cohort = object@cohort[!is_placebo]),
+      "There must be only one dose level, other than placebo, per cohort"
+    )
   }
```

Comment:
@danielinteractive I think this above check should also be conduced when `object@placebo` is `TRUE`, that is why I placed it out of `if-else` clause.